### DISCUSSION TITLE
Simple TS implementation

### DISCUSSION THREAD

**Discussion by taikichi-pc** [meta-pytorch/botorch#1681] [Q&A] [answered]
Thank you for your previous kind reply.

I'd like to implement TS(Thompson sampling) baseline for comparing with other acquisition function.

**How would you implement TS using botorch as a simple(basic) code?**

I've already read bellow links.
https://botorch.org/tutorials/turbo_1
https://botorch.org/tutorials/thompson_sampling

BTW COMBO, another BO library, and so on adopt TS as a default algorithm for Multi-objective.
I think It's classical and good(especially in terms of calc. cost) way and not complicated.
But I look forward to refer your code.

Thank you.

**Comment by Balandat** [comment, marked as answer]
Take a look at [`MaxPosteriorSampling`](https://github.com/pytorch/botorch/blob/3e18d2ab45f839dcb82f0436d8100f0716716f5c/botorch/generation/sampling.py#L66-L176) - this is Thompson sampling. The support for batched models makes it a bit less readable and there a some tensor acrobatics necessary if you want to avoid duplicated samples in that setting, but overall it's pretty straightforward. 

Note that this draws samples at some discrete set of points `X`. Another implementation would be to draw sample paths from the model posterior (using the pathwise sampling feature added in #1463), and then optimize each of them individually using some gradient-based scheme. This is currently not (yet) implemented though.
